components/renku_data_services/notebooks/api/classes/k8s_client.py

```diff
@@ -534,27 +678,266 @@ async def delete_server(self, server_name: str, safe_username: str) -> None:
+    ) -> StatefulSet | None:
+        return await self._k8s_client.patch_statefulset(server_name, patch)
 
     async def create_secret(self, secret: V1Secret) -> V1Secret:
-        """Create a secret."""
-        return await self.renku_ns_client.create_secret(secret)
+        return await self._k8s_client.create_secret(secret)
 
     async def delete_secret(self, name: str) -> None:
-        """Delete a secret."""
-        return await self.renku_ns_client.delete_secret(name)
+        await self._k8s_client.delete_secret(name)
 
-    async def patch_secret(self, name: str, patch: dict[str, Any] | list[dict[str, Any]]) -> None:
-        """Patch a secret."""
-        return await self.renku_ns_client.patch_secret(name, patch)
+
+class MultipleK8sClient(K8sClientProto[_SessionType, _Kr8sType]):
+    """Multiple Kubernetes cluster client wrapper."""
+
+    def __init__(
+        self,
+        server_type: type[_SessionType],
+        kr8s_type: type[_Kr8sType],
+        cache_url: str,
+        username_label: str,
+        rp_repo: ResourcePoolRepository,
+        skip_cache_if_unavailable: bool = False,
+    ):
+        self._clients: dict[str, _SingleK8sClient[_SessionType, _Kr8sType]] = dict()
+        self._kube_conf_root_dir = "/secrets/kube_configs"
+        self._rp_repo = rp_repo
+
+        # Add at least one connection, to the default cluster.
+        self._clients[DEFAULT_K8S_CLUSTER] = _SingleK8sClient(
+            server_type,
+            kr8s_type,
+            kr8s.api(),
+            cache_url,
+            username_label,
+            skip_cache_if_unavailable,
+        )
+
+        if os.path.exists(self._kube_conf_root_dir):
+            for f in os.scandir(self.kube_conf_root_dir):
+                self._clients[f.name.removesuffix(".yaml")] = _SingleK8sClient(
+                    server_type,
+                    kr8s_type,
+                    kr8s.api(kubeconfig=f"{self.kube_conf_root_dir}/{f.name}"),
+                    cache_url,
+                    username_label,
+                    skip_cache_if_unavailable,
+                )
+        else:
+            logging.warning(f"Cannot open directory '{self._kube_conf_root_dir}', ignoring kube configs...")
+
+        # maps session/server name to k8s client
+        self._session2client: dict[str, _SingleK8sClient[_SessionType, _Kr8sType]] = {}
+
+    async def _client_by_session(
+        self, session_name: str, safe_username: str
+    ) -> _SingleK8sClient[_SessionType, _Kr8sType] | None:
+        # Try in our cache, if not there look it up in K8s and update our cache
+        client = self._session2client.get(session_name, None)
+        if client is None:
+            for c in self._clients.values():
+                if session_name in await c.list_servers(safe_username):
+                    self._session2client[session_name] = c
+                    client = c
+                    break
```

Comment:
Since we run with multiple instances of the data service we will not have the complete picture in the server name to client cache. So that means you will hit this part of the code often.

I think in a followup PR we should tackle some of these problems. The main one here and in a few other places in the code is that right now we assume that the cached client cannot:
- give you all servers across all clusters for a user
- efficiently find a cluster name by the server name

But it will be able to do both all the time in an efficient way. Anyhow you do not have to do anything for this comment. I just wanted to mention this to discuss in the next sync for the build.